GenericWhitespaceCheck: Handling of whitespace between generic and record header

From https://github.com/checkstyle/checkstyle/pull/14497#discussion_r1493346388 (second last paragraph)

([Check Documentation](https://checkstyle.sourceforge.io/checks/whitespace/genericwhitespace.html))

**Goal of this Issue:**
- Conclude whether whitespace between generic and record header should be considered as a violation or not. 
- Based on the conclusion necessary changes can be done.

**How it works Now:**
Currently whitespace between generic and record header is not considered a violation.

**Test.java**
```java
record Test<T> () {}
//            ^
//            +--- whitespace in question
```

**config.xml**
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="GenericWhitespace"/>
  </module>
</module>
```

**Current CLI Execution:**
```
java -jar checkstyle-10.13.0-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```

**Expected Working:** ~~To be decided~~
`[ERROR] Test.java:1:14: '>' is followed by whitespace. [GenericWhitespace]`

### Findings that may be useful:
- The syntax is kind of similar to Issue [#14344](https://github.com/checkstyle/checkstyle/issues/14344)
- [Horizontal Whitespace](https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace) from google style, which includes the *only* places where occurrence of whitespace is valid does not includes this usage.
- In [oracle docs](https://docs.oracle.com/en/java/javase/17/language/records.html#GUID-6699E26F-4A9B-4393-A08B-1E47D4B2D263) for record classes [see first point of **Features of Record Classes**], a whitespace *is* present between generic and record header in the example.
- This check adheres to the "typical convention".

**Typical Usages of Generic Records (Anecdotal):**
- Generic records in Checkstyle, [default projects](https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/projects-to-test-on-for-github-action.properties) for regression ,and those I have encountered so far, do not contain whitespace between generic and record header.
This can be verified by running the following two grep cmds:
  1. `grep -P -r "record [a-zA-Z0-9\s$_]+\<.*\>\(.*\)" dir/ | tee >(wc -l)`
  2. `grep -P -r "record [a-zA-Z0-9\s$_]+\<.*\>\s+\(.*\)" dir/ | tee >(wc -l)`
- The pattern will match for the generic records, first one matching no-whitespace between generic and record header, while the second one matching atleast a single-whitespace between generic and record header.
- Output for the cmd on Checkstyle repo: **i.** `32`; **ii:** `0`
- Output for the cmd on regression repos: **i.** `60`; **ii.** `0`
**Side note:** Arguably better way to verify that there are no such cases would be implement changes which detects space between generic and record header as a violation, and then check for differences in the report. (lmk if that is needed)

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java
@@ -51,7 +51,8 @@ import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
  * <ul>
  * <li> should not be preceded with whitespace in all cases.</li>
  * <li> should be followed with whitespace in almost all cases,
- *   except diamond operators and when preceding method name or constructor.</li></ul>
+ *   except diamond operators and when preceding a method name, constructor, or record header.</li>
+ * </ul>
  * <p>
  * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
  * </p>
@@ -219,7 +220,9 @@ public class GenericWhitespaceCheck extends AbstractCheck {
      */
     private void processSingleGeneric(DetailAST ast, int[] line, int after) {
         final char charAfter = Character.toChars(line[after])[0];
-        if (isGenericBeforeMethod(ast) || isGenericBeforeCtorInvocation(ast)) {
+        if (isGenericBeforeMethod(ast)
+                || isGenericBeforeCtorInvocation(ast)
+                || isGenericBeforeRecordHeader(ast)) {
             if (Character.isWhitespace(charAfter)) {
                 log(ast, MSG_WS_FOLLOWED, CLOSE_ANGLE_BRACKET);
             }
@@ -227,6 +230,22 @@ public class GenericWhitespaceCheck extends AbstractCheck {
         else if (!isCharacterValidAfterGenericEnd(charAfter)) {
             log(ast, MSG_WS_ILLEGAL_FOLLOW, CLOSE_ANGLE_BRACKET);
         }
+    }
+
+    /**
+     * Checks if generic is before record header. Identifies two cases:
+     * <ol>
+     *     <li>In record def, eg: {@code record Session<T>()}</li>
+     *     <li>In record pattern def, eg: {@code o instanceof Session<String>(var s)}</li>
+     * </ol>
+     *
+     * @param ast ast
+     * @return true if generic is before record header
+     */
+    private static boolean isGenericBeforeRecordHeader(DetailAST ast) {
+        final DetailAST grandParent = ast.getParent().getParent();
+        return grandParent.getType() == TokenTypes.RECORD_DEF
+                || grandParent.getParent().getType() == TokenTypes.RECORD_PATTERN_DEF;
     }
 
     /**
@@ -370,10 +389,9 @@ public class GenericWhitespaceCheck extends AbstractCheck {
      * @return checks if given character is valid
      */
     private static boolean isCharacterValidAfterGenericEnd(char charAfter) {
-        return charAfter == '(' || charAfter == ')'
-            || charAfter == ',' || charAfter == '['
-            || charAfter == '.' || charAfter == ':'
-            || charAfter == ';'
+        return charAfter == ')' || charAfter == ','
+            || charAfter == '[' || charAfter == '.'
+            || charAfter == ':' || charAfter == ';'
             || Character.isWhitespace(charAfter);
     }
 

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/whitespace/GenericWhitespaceCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/whitespace/GenericWhitespaceCheck.xml
@@ -26,7 +26,8 @@
  &lt;ul&gt;
  &lt;li&gt; should not be preceded with whitespace in all cases.&lt;/li&gt;
  &lt;li&gt; should be followed with whitespace in almost all cases,
-   except diamond operators and when preceding method name or constructor.&lt;/li&gt;&lt;/ul&gt;</description>
+   except diamond operators and when preceding a method name, constructor, or record header.&lt;/li&gt;
+ &lt;/ul&gt;</description>
          <message-keys>
             <message-key key="ws.followed"/>
             <message-key key="ws.illegalFollow"/>

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheckTest.java
@@ -228,7 +228,45 @@ public class GenericWhitespaceCheckTest
 
     @Test
     public void testBeforeRecordHeader() throws Exception {
-        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        final String[] expected = {
+            "17:20: " + getCheckMessage(MSG_WS_FOLLOWED, '>'),
+            "18:20: " + getCheckMessage(MSG_WS_FOLLOWED, '<'),
+            "18:20: " + getCheckMessage(MSG_WS_PRECEDED, '<'),
+            "18:24: " + getCheckMessage(MSG_WS_FOLLOWED, '>'),
+            "18:24: " + getCheckMessage(MSG_WS_PRECEDED, '>'),
+            "30:27: " + getCheckMessage(MSG_WS_FOLLOWED, '>'),
+            "30:38: " + getCheckMessage(MSG_WS_PRECEDED, '<'),
+            "30:80: " + getCheckMessage(MSG_WS_ILLEGAL_FOLLOW, '>'),
+            "36:38: " + getCheckMessage(MSG_WS_FOLLOWED, '>'),
+            "43:44: " + getCheckMessage(MSG_WS_PRECEDED, '<'),
+            "43:69: " + getCheckMessage(MSG_WS_FOLLOWED, '>'),
+            "49:21: " + getCheckMessage(MSG_WS_PRECEDED, '<'),
+            "49:64: " + getCheckMessage(MSG_WS_PRECEDED, '>'),
+            "49:66: " + getCheckMessage(MSG_WS_FOLLOWED, '>'),
+            "56:63: " + getCheckMessage(MSG_WS_PRECEDED, '<'),
+            "56:80: " + getCheckMessage(MSG_WS_FOLLOWED, '>'),
+            "62:36: " + getCheckMessage(MSG_WS_PRECEDED, '<'),
+            "62:61: " + getCheckMessage(MSG_WS_FOLLOWED, '>'),
+            "67:49: " + getCheckMessage(MSG_WS_FOLLOWED, '>'),
+            "73:26: " + getCheckMessage(MSG_WS_PRECEDED, '<'),
+            "73:51: " + getCheckMessage(MSG_WS_FOLLOWED, '>'),
+            "73:64: " + getCheckMessage(MSG_WS_FOLLOWED, '<'),
+            "80:26: " + getCheckMessage(MSG_WS_PRECEDED, '<'),
+            "80:34: " + getCheckMessage(MSG_WS_FOLLOWED, '<'),
+            "80:55: " + getCheckMessage(MSG_WS_ILLEGAL_FOLLOW, '>'),
+            "91:25: " + getCheckMessage(MSG_WS_FOLLOWED, '>'),
+            "91:44: " + getCheckMessage(MSG_WS_PRECEDED, '<'),
+            "91:47: " + getCheckMessage(MSG_WS_PRECEDED, '>'),
+            "91:61: " + getCheckMessage(MSG_WS_PRECEDED, '<'),
+            "91:71: " + getCheckMessage(MSG_WS_FOLLOWED, '>'),
+            "91:73: " + getCheckMessage(MSG_WS_PRECEDED, '>'),
+            "101:25: " + getCheckMessage(MSG_WS_PRECEDED, '<'),
+            "101:58: " + getCheckMessage(MSG_WS_FOLLOWED, '>'),
+            "108:25: " + getCheckMessage(MSG_WS_PRECEDED, '<'),
+            "108:32: " + getCheckMessage(MSG_WS_FOLLOWED, '>'),
+            "108:46: " + getCheckMessage(MSG_WS_FOLLOWED, '<'),
+            "108:63: " + getCheckMessage(MSG_WS_FOLLOWED, '>'),
+        };
         verifyWithInlineConfigParser(
                 getNonCompilablePath("InputGenericWhitespaceBeforeRecordHeader.java"),
                 expected);

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/genericwhitespace/InputGenericWhitespaceBeforeRecordHeader.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/genericwhitespace/InputGenericWhitespaceBeforeRecordHeader.java
@@ -4,9 +4,116 @@ GenericWhitespace
 
 */
 
-//non-compiled with javac: Compilable with Java16
+//non-compiled with javac: Compilable with Java21
 package com.puppycrawl.tools.checkstyle.checks.whitespace.genericwhitespace;
 
+import java.lang.annotation.Target;
+import java.util.HashMap;
+
+import static java.lang.annotation.ElementType.TYPE_USE;
+
 public class InputGenericWhitespaceBeforeRecordHeader {
-    record Session<T>() {}
+    record License<T>() {}
+    record Person<T> () {} // violation, ''>' is followed by whitespace.'
+    record Session < T > (T a) {}
+    // 4 violations above:
+    //                    ''<' is followed by whitespace.'
+    //                    ''<' is preceded with whitespace.'
+    //                    ''>' is followed by whitespace.'
+    //                    ''>' is preceded with whitespace.'
+
+    class InnerClass {
+        static <T> void innerMethod(T a) {}
+        static <T> void innerMethod(T a, boolean isSession) {}
+        record Participant<T>() {
+            //            +          +                                         +
+            record Stats<T> (HashMap <String, Session<HashMap<String, Integer>>>stats) {
+                // 3 violations above:
+                //                    ''>' is followed by whitespace.'
+                //                    ''<' is preceded with whitespace.'
+                //                    ''>' is followed by an illegal character.'
+
+                record Achievements<T> () {} // violation, ''>' is followed by whitespace.'
+            }
+        }
+    }
+
+    int method(Session<HashMap<String, Integer>> ssn) {
+        //                                 +                        +
+        InnerClass.innerMethod(new Session <HashMap<Integer, String>> (new HashMap()));
+        // 2 violations above:
+        //                    ''<' is preceded with whitespace.'
+        //                    ''>' is followed by whitespace.'
+
+        //          +                                          + +
+        new License <@A Session<HashMap<@A Session, @A Session >>> ();
+        // 3 violations above:
+        //                    ''<' is preceded with whitespace.'
+        //                    ''>' is preceded with whitespace.'
+        //                    ''>' is followed by whitespace.'
+
+        //                                                    +                +
+        new InnerClass.Participant.Stats.Achievements<Session <License<String>>> ();
+        // 2 violations above:
+        //                   ''<' is preceded with whitespace.'
+        //                   ''>' is followed by whitespace.'
+
+        //                         +                        +
+        if (ssn instanceof Session <HashMap<String, Integer>> (var sess)) {
+            // 2 violations above:
+            //                    ''<' is preceded with whitespace.'
+            //                    ''>' is followed by whitespace.'
+
+            InnerClass.innerMethod(new Session<?> []{}, true);
+            // violation above, ''>' is followed by whitespace.'
+        }
+
+        return switch (ssn) {
+            //           +                        +            +
+            case Session <HashMap<String, Integer>> (@A HashMap< @A String, Integer> hm) -> 1;
+            //  3 violations above:
+            //                     ''<' is preceded with whitespace.'
+            //                     ''>' is followed by whitespace.'
+            //                     ''<' is followed by whitespace.'
+
+            //           +       +                    +
+            case Session <HashMap< @A String, Integer>>sess -> 1;
+            // 3 violations above:
+            //                    ''<' is preceded with whitespace.'
+            //                    ''<' is followed by whitespace.'
+            //                    ''>' is followed by an illegal character.'
+        };
+    }
+
+    static {
+        new Object() {
+            //          +                  +  +             +         + +
+            record Sys<T> (Session<Session <T >> p, Session <Session<T> > q) {}
+            // 6 violations above:
+            //                    ''>' is followed by whitespace.'
+            //                    ''<' is preceded with whitespace.'
+            //                    ''>' is preceded with whitespace.'
+            //                    ''<' is preceded with whitespace.'
+            //                    ''>' is followed by whitespace.'
+            //                    ''>' is preceded with whitespace.'
+
+            //          +                                +
+            record Game <T extends Number & Comparable<T>> (T guess) {}
+            // 2 violations above:
+            //                    ''<' is preceded with whitespace.'
+            //                    ''>' is followed by whitespace.'
+
+            void method() {
+                //      +      +             +                +
+                new Sys <String> (new Session< Session<String>> (new Session<>("a")), null);
+                // 4 violations above:
+                //                    ''<' is preceded with whitespace.'
+                //                    ''>' is followed by whitespace.'
+                //                    ''<' is followed by whitespace.'
+                //                    ''>' is followed by whitespace.'
+            }
+        };
+    }
+
+    @Target({TYPE_USE}) @interface A {}
 }

--- a/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/genericwhitespace/Example1.txt
+++ b/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/genericwhitespace/Example1.txt
@@ -23,4 +23,6 @@ List<T> list = ImmutableList.Builder<T>::new;
 sort(list, Comparable::<String>compareTo);
 // Constructor call
 MyClass obj = new <String>MyClass();
+// Record header
+record License<T>() {}
 // xdoc section -- end

--- a/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/genericwhitespace/Example2.txt
+++ b/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/genericwhitespace/Example2.txt
@@ -14,4 +14,6 @@ public<T> void foo() {} // violation, "<" not preceded with whitespace
 List a = new ArrayList<> (); // violation, ">" followed by whitespace
 Map<Integer, String>m; // violation, ">" not followed by whitespace
 Pair<Integer, Integer > p; // violation, ">" preceded with whitespace
+
+record License<T> () {} // violation, ">" followed by whitespace
 // xdoc section -- end

--- a/src/xdocs/checks/whitespace/genericwhitespace.xml
+++ b/src/xdocs/checks/whitespace/genericwhitespace.xml
@@ -30,7 +30,7 @@
         <ul>
           <li> should not be preceded with whitespace in all cases.</li>
           <li> should be followed with whitespace in almost all cases, except diamond operators
-               and when preceding method name or constructor.</li>
+            and when preceding a method name, constructor, or record header.</li>
         </ul>
       </subsection>
 
@@ -65,6 +65,8 @@ List&lt;T&gt; list = ImmutableList.Builder&lt;T&gt;::new;
 sort(list, Comparable::&lt;String&gt;compareTo);
 // Constructor call
 MyClass obj = new &lt;String&gt;MyClass();
+// Record header
+record License&lt;T&gt;() {}
         </source>
         <p id="Example2-code">
           Examples with incorrect spacing:
@@ -77,6 +79,8 @@ public&lt;T&gt; void foo() {} // violation, &quot;&lt;&quot; not preceded with w
 List a = new ArrayList&lt;&gt; (); // violation, &quot;&gt;&quot; followed by whitespace
 Map&lt;Integer, String&gt;m; // violation, &quot;&gt;&quot; not followed by whitespace
 Pair&lt;Integer, Integer &gt; p; // violation, &quot;&gt;&quot; preceded with whitespace
+
+record License&lt;T&gt; () {} // violation, &quot;&gt;&quot; followed by whitespace
         </source>
       </subsection>
 

--- a/src/xdocs/checks/whitespace/genericwhitespace.xml.template
+++ b/src/xdocs/checks/whitespace/genericwhitespace.xml.template
@@ -30,7 +30,7 @@
         <ul>
           <li> should not be preceded with whitespace in all cases.</li>
           <li> should be followed with whitespace in almost all cases, except diamond operators
-               and when preceding method name or constructor.</li>
+            and when preceding a method name, constructor, or record header.</li>
         </ul>
       </subsection>
 

